middleware/request_id.go

```diff
@@ -86,3 +86,22 @@ func GetReqID(ctx context.Context) string {
 func NextRequestID() uint64 {
 	return atomic.AddUint64(&reqid, 1)
 }
+
+// RequestIDFromHeader is a middleware similar to RequestID
+// except it uses the value from request header X-Request-Id
+// if one is available. Else revert to default behavior.
+func RequestIDFromHeader(next http.Handler) http.Handler {
```

Comment:
Can we provide a `SetReqID()` function, instead of hardcoding the X-Request-Id header name?

Reply:
Would be nice to provide and example for #368 as part of this PR.

Reply:
I thought about this first. However I thought it was more intrusive change. The RequestID is maintained in a context variable keyed by a non-exported variable. Exporting that variable doesn't sound good either. So it would have a SetReqID() method inside the middleware which also takes the context as an explicit parameter, and set the given string as the request-id. Is that what you are looking for?  As for #368 , isn't there already a `GetReqID()` method available in the middleware?

Reply:
You know what.. I think it's not a bad idea to use the `X-Request-ID` header directly after all. This middleware is opinionated anyway.

Imho, the is the desired lifecycle of "tracing" the request-id across multiple services is as follows:

1. browser request

 ->

2. optionally, an external service, creates reqID and passes it as X-Request-ID header

 ->

3. our program reads the header and sets the context value, ie.
```go
r.Use(middleware.RequestID)

// which behind the scenes does this:
reqID := r.Header.Get("X-Request-ID")
if reqID == "" {
  reqID = generateReqId()
}
setReqId(ctx, reqID)
```

 ->

4. middlewares/handlers use the reqID value (from ctx) and pass it to logger or to external HTTP calls as X-Request-ID header

    we just provide a getter for this:
```go
reqID := middleware.GetReqId(ctx)

// log.With().String("requestId", reqID) // zerolog structured log example
// req.Header.Set("X-Request-ID", reqID) // HTTP calls to other services that should be aware of this request ID
```

Reply:
Item 4 already exists. Please confirm if that is what you meant.
So the remaining change is to merge RequestIDFromHeader in to RequestID middleware. I wasn't sure if it would break any existing usage.